### DISCUSSION TITLE
Airflow UI DAG Import Errors

### DISCUSSION THREAD

**Discussion by bparhy** [apache/airflow#20793] [Q&A]
### Apache Airflow version

2.1.3

### What happened

We are using airflow in k8s. Webserver and scheduler are running without any errors in their logs.

Below is the snippet of a DAG.

from airflow import DAG
from airflow.operators.python_operator import PythonOperator
from airflow.providers.snowflake.hooks.snowflake import SnowflakeHook
from airflow.providers.snowflake.operators.snowflake import SnowflakeOperator
from bi_plugin import SnowflakeMultiSqlStatmentOperator

I am seeing the below error in the UI. But the dag is parsing perfect and I can see its code and I can run it with out any issue.

Broken DAG: [/usr/local/airflow/dags/network/vbc_telemetry_feed_prod.py] Traceback (most recent call last):
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/usr/local/airflow/dags/network/vbc_telemetry_feed_prod.py", line 5, in <module>
    from bi_plugin import SnowflakeMultiSqlStatmentOperator
ModuleNotFoundError: No module named 'bi_plugin'

We used to have plugin code in this bi_plugin module with 2.1.3 migration we are still keeping the bi_plugin file in the plugins folder which is again inside the dags folder and our imports are changed to 

from bi_plugin import SnowflakeMultiSqlStatmentOperator

Please let me know. Any help in this direction is appreciated.

### What you expected to happen

I was expecting no errors in the UI.

### Operating System

linux

### Deployment

Other 3rd-party Helm chart

### Deployment details

This is a kubernetes deployment with two schedulers and one webserver.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by kaxil** [comment]
Converting it to a GitHub discussion as it is not a bug and mostly likely related to pythonpath's and import.

Where it the `bi_plugin.py` file stored?

**Reply by bparhy** [reply]
Thanks @kaxil for your quick response. bi_plugin.py is located at
/usr/local/airflow/dags/plugins
I am just seeing this error in the UI. I can see the DAGs in UI and they are parsing fine.

**Reply by potiuk** [reply]
Generally speaking "plugins" should not be used to import files from dags. It should not be used to import files from there.

You should restructure your DAG code to import files from the PYTHONPATH following this guideline: 

https://airflow.apache.org/docs/apache-airflow/stable/modules_management.html

**Reply by potiuk** [reply]
Simply - whenever you import something your "package path" starts where PYTHONPATH is (see the docs I lined to for explanation on what is the PYTHONPATH setting by default.)

**Reply by potiuk** [reply]
In your case (if you sitll want to keep the plugins folder) it should be `from plugins.bi_plugin import ...."  (but you have to make sure __init__.py is created in plugins folder.

Just follow the usual rules that are there for importing python code.

**Reply by bparhy** [reply]
ok thanks testing.

**Comment by bparhy** [comment]
Thanks @potiuk I tested but still see the same error where the UI is complaining about the dag parsing error but If I go the dag its parsing correct and I am able to run it. 

However if I change the replicas from 2 to 1 for the scheduler this error is going away. I dont see any error in the UI and all the dags are parsing.
Please let me know if this is related.

**Reply by potiuk** [reply]
Likely the second replica misses some volumes mounted.

**Reply by potiuk** [reply]
Or has no access.

**Reply by potiuk** [reply]
Or is misconfigured.

**Reply by bparhy** [reply]
thanks @potiuk  I have the same exact deployment helm chart for scheduler only change is with 2 schedulers I have **replica** set to 2.

**Reply by potiuk** [reply]
Then likely it is a bug in your chart. If this is not the Official chart of Apache Airflow (you indicated it is some other) then you should likely raise to whowever created that chart. Unfortutnately it's hard for me to say what is in the chart and the chart author will be the best to investigate it

**Reply by potiuk** [reply]
Or switch to the official chart of Airflow. It is tested and works with multiple schedulers and when you have problems like this, the community will be able to help you. See https://airflow.apache.org/docs/helm-chart/stable/index.html
